x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/Concat.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.eql.expression.function.scalar.string;
+
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.Expressions;
+import org.elasticsearch.xpack.ql.expression.Expressions.ParamOrdinal;
+import org.elasticsearch.xpack.ql.expression.function.scalar.ScalarFunction;
+import org.elasticsearch.xpack.ql.expression.gen.pipeline.Pipe;
+import org.elasticsearch.xpack.ql.expression.gen.script.ParamsBuilder;
+import org.elasticsearch.xpack.ql.expression.gen.script.ScriptTemplate;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.StringJoiner;
+
+import static org.elasticsearch.xpack.eql.expression.function.scalar.string.ConcatFunctionProcessor.doProcess;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isExact;
+import static org.elasticsearch.xpack.ql.expression.gen.script.ParamsBuilder.paramsBuilder;
+
+/**
+ * EQL specific concat function to build a string of all input arguments concatenated.
+ */
+public class Concat extends ScalarFunction {
+
+    private final List<Expression> values;
+
+    public Concat(Source source, List<Expression> values) {
+        super(source, values);
+        this.values = values;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (!childrenResolved()) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = TypeResolution.TYPE_RESOLVED;
+        int index = 0;
+        for (Expression value : values) {
+            resolution = isExact(value, sourceText(), ParamOrdinal.fromIndex(index));
```

Comment:
Why only `isExact`? Shouldn't this be a string only type of value?

Reply:
Also, this loop goes through all of `concat`'s values and returns the resolution of the last element in list. Shouldn't, maybe, stop at the first resolution that is `.unresolved()` and return that?

Reply:
fixed in 9295e7920777ebca4c3d26cc51cedecdbf4869db

Reply:
also, this accepts non string parameters as well, and calls `.toString()` on them
the example uses `pid` which is a long
https://eql.readthedocs.io/en/latest/query-guide/functions.html#concat